sessions/defense-mechanisms/README.md

```diff
@@ -0,0 +1,423 @@
+
+Let's start by deactivating ASLR, which is going to be discussed in the following section of this tutorial, and only focus on the NX protection. We can do this in two ways, as told below.
+
+    *   To disable ASLR system-wide we use (root access is required): `sudo bash -c 'echo 0 > /proc/sys/kernel/randomize_va_space'`;
+    *   To create a shell with ASLR disabled (ASLR will also be disabled for future processes spawned from that shell), we use (root access is not required): `setarch $(uname -m) -R /bin/bash`.
+
+After disabling ASLR, let's compile an extremely simple C application. Save the following code as `hello.c`:
+
+```C
+int main() {
+    while (1);
+}
+```
+
+Make sure you have both `build-essential` and `gcc-multilib` packages installed before going further (run `sudo apt install build-essential gcc-multilib` on Debian-based systems).
+
+Compile the `hello.c` code using `CFLAGS='-m32 -O0' make hello`. The result should be a `hello` binary.
+
+As presented in the `Static Analysis` session, the ELF format contains flags for each segment that specify what permissions should be granted. You can use `readelf -l hello` to dump all program headers for this binary. The result should be similar to:
+
+```
+Program Headers:
+  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
+  PHDR           0x000034 0x08048034 0x08048034 0x00120 0x00120 R E 0x4
+  INTERP         0x000154 0x08048154 0x08048154 0x00013 0x00013 R   0x1
+      [Requesting program interpreter: /lib/ld-linux.so.2]
+  LOAD           0x000000 0x08048000 0x08048000 0x00568 0x00568 R E 0x1000
+  LOAD           0x000f08 0x08049f08 0x08049f08 0x00114 0x00118 RW  0x1000
+  DYNAMIC        0x000f14 0x08049f14 0x08049f14 0x000e8 0x000e8 RW  0x4
+  NOTE           0x000168 0x08048168 0x08048168 0x00044 0x00044 R   0x4
+  GNU_EH_FRAME   0x000490 0x08048490 0x08048490 0x0002c 0x0002c R   0x4
+  GNU_STACK      0x000000 0x00000000 0x00000000 0x00000 0x00000 RW  0x10
+  GNU_RELRO      0x000f08 0x08049f08 0x08049f08 0x000f8 0x000f8 R   0x1
+
+ Section to Segment mapping:
+  Segment Sections...
+   00     
+   01     .interp 
+   02     .interp .note.ABI-tag .note.gnu.build-id .gnu.hash .dynsym .dynstr .gnu.version .gnu.version_r .rel.dyn .rel.plt .init .plt .text .fini .rodata .eh_frame_hdr .eh_frame 
+   03     .init_array .fini_array .jcr .dynamic .got .got.plt .data .bss 
+   04     .dynamic 
+   05     .note.ABI-tag .note.gnu.build-id 
+   06     .eh_frame_hdr 
+   07     
+   08     .init_array .fini_array .jcr .dynamic .got
+```
+
+Check the `Flg` column. For example, the first `LOAD` segment contains `.text` and is marked `R E`, while the `GNU_STACK` segment is marked `RW `.
+
+Next we are interested in seeing calls to `mmap2()` and `mprotect()` made by the loader. We are going to use the `strace` tool for this, and directly execute the loader. You can check the path to the loader on your system using `ldd hello`.
+
```

Comment:
I'd suggest changing `ldd hello` to something like `ldd /bin/ls`

Reply:
The `hello` is used throughout the lab. Still, does it add any benefit if we interchange the two?

Reply:
In my opinion, it would be easier to remember for later use, but that may not be true for everyone

Reply:
The thing with `hello` is that it seen as a point of interest throughout the lab. You start from the source file `hello.c`, you then compile `hello` from `hello.c`, then start inspecting the binary (with `readelf`, `strace`, etc.). So it kind of feels natural to use it as an argument for `ldd` as well.